zigen-project/zen
# add popup damage

## Context

#116, #122 

## Summary

- Added xdg_shell's popup damage.
- This allows, for example, right-clicking on a wayland client to display/hide the popup. Details of what can now be done are added in the "How to check behavior" section.
- Added view_child struct to handle damage.

## How to check behavior
The following behaviors can be checked with the Wayland clients (e.g., `weston-terminal` `weston-flower`)
- showing popup
  - When you right click on the view, a popup appears around the cursor.
    - The requirement that if the cursor is around the edge of the screen, the popup appear inside the screen, is todo.
  - When you right click anywhere else, the previous popup disappears and a new popup appears around the cursor.
- hiding popup
  - (right/left) click elsewhere in the view
  - close the view
  - start `move`
  - focus/restack another view

- The behavior related to board is not defined.
  - When the view is focused during the board operation, the popup will remain.
  - When the view is unfocused during the board operation, the popup will disappear.

NOTE:
Below are the TODOs.
- popup's popup (popup tree)
  - handle the case that view_child's parent(view_child) exists.
- if the cursor is around the edge of the screen, the popup appear inside the screen.
- (subsurface)
